File: EasyILive_iOS/EVRTCRootViewController.m
//
//  EVRTCRootViewController.m
//  EVRTC
//
//  Created by Lcrnice on 2017/7/28.
//  Copyright © 2017年 Easyvaas. All rights reserved.
//

#import "EVRTCRootViewController.h"
#import "CCAlertManager.h"
#import "EVMediaAuth.h"
#import <EVSDKBaseFramework/EVSDKManager.h>
#import "EVRTCViewController.h"
#import "EVScanQRViewController.h"
#import "EVRTCSettingViewController.h"

#define CHANNEL_ACCEPTABLE_CHARACTERS @"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
#define UID_ACCEPTABLE_CHARACTERS @"0123456789"
static NSString * const kShowRTC = @"ShowRTC";

@interface EVRTCRootViewController () <UITextFieldDelegate>
@property (weak, nonatomic) IBOutlet UILabel *versionLabel;
@property (weak, nonatomic) IBOutlet UITextField *channelTF;
@property (weak, nonatomic) IBOutlet UITextField *uidTF;

@property (nonatomic, assign) EVRtcVideoProfile currentProfile;

@end

@implementation EVRTCRootViewController

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self addObserver];
    [EVSDKManager initSDKWithAppID:@"" appKey:@"" appSecret:@"" userID:@"rtcTester"];
    
    self.currentProfile = EVRtcVideoProfile_640x360;
    self.versionLabel.text = [NSString stringWithFormat:@"Versoin:%@", [EVSDKManager SDKVersion]];
    self.channelTF.delegate = self;
    self.uidTF.delegate = self;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - regist SDK
- (void)addObserver{
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(initSDKError:) name:EVSDKInitErrorNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(initSDKSuccess) name:EVSDKInitSuccessNotification object:nil];
}

- (void)initSDKError:(NSNotification *)notification{
    NSLog(@"---notification:%@", notification);
    NSString *message = [NSString stringWithFormat:@"%@", notification.object];
    [[CCAlertManager shareInstance] performComfirmTitle:notification.name message:message comfirmTitle:@"确定" WithComfirm:nil];
}

- (void)initSDKSuccess{
    NSLog(@"SDK 初始化成功");
}

#pragma mark - actions
- (IBAction)unwindSeguToRootViewController:(UIStoryboardSegue *)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}
- (IBAction)unwindSeguConfigToRootViewController:(UIStoryboardSegue *)sender {
    UIViewController *sourceVC = sender.sourceViewController;
    
    if ([sourceVC isKindOfClass:[EVRTCSettingViewController class]]) {
        EVRTCSettingViewController *settingVC = (EVRTCSettingViewController *)sourceVC;
        self.currentProfile = settingVC.currentProfile;
    }
    
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)scanQR:(id)sender {
    EVScanQRViewController *scanVC = [EVScanQRViewController new];
    __weak typeof(self) wSelf = self;
    scanVC.getQrCode = ^(NSString *string) {
        __strong typeof(wSelf) sSelf = wSelf;
        if (string.length > 0) {
            sSelf.channelTF.text = string;
        }
    };
    [self presentViewController:scanVC animated:YES completion:nil];
}

- (IBAction)joinChannel:(id)sender {
    if (![self p_isValidSDK]) {
        return;
    }
    
    if (![self p_isValidChannel]) {
        return;
    }
    
    [EVMediaAuth checkAndRequestMicPhoneAndCameraUserAuthed:^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self showDetailVCWithRole:EVRtc_ClientRole_LiveGuest];
        });
    } userDeny:nil];
}

- (IBAction)createChannel:(id)sender {
    if (![self p_isValidSDK]) {
        return;
    }
    
    [EVMediaAuth checkAndRequestMicPhoneAndCameraUserAuthed:^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self showDetailVCWithRole:EVRtc_ClientRole_Master];
        });
    } userDeny:nil];
}

- (IBAction)watchLiveInChannel:(id)sender {
    if (![self p_isValidSDK]) {
        return;
    }
    
    if (![self p_isValidChannel]) {
        return;
    }
    
    [self showDetailVCWithRole:EVRtc_ClientRole_Guest];
}

- (IBAction)easyvass:(id)sender {
    NSURL *url = [NSURL URLWithString:@"http://easyvaas.com"];
    if ([[UIApplication sharedApplication] canOpenURL:url]) {
        [[UIApplication sharedApplication] openURL:url];
    }
}

- (void)showDetailVCWithRole:(EVRtcClientRole)role {
    [self performSegueWithIdentifier:kShowRTC sender:@(role)];
}


#pragma mark - Navigation
// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
    NSString *identifier = segue.identifier;
    [self resignTextFileds];
    
    if ([identifier isEqualToString:kShowRTC]) {
        EVRTCViewController *rtcVC = segue.destinationViewController;
        rtcVC.roomName = self.channelTF.text;
        rtcVC.uidString = self.uidTF.text;
        rtcVC.role = [sender integerValue];
        rtcVC.profile = self.currentProfile;
    }
    
    if ([segue.destinationViewController isKindOfClass:[UINavigationController class]]) {
        UINavigationController *nav = segue.destinationViewController;;
        EVRTCSettingViewController *settingVC = (EVRTCSettingViewController *)[nav.viewControllers firstObject];
        settingVC.currentProfile = self.currentProfile;
    }
    
}

#pragma mark - Delegates
- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string  {
    if (textField == self.channelTF) {
        NSCharacterSet *cs = [[NSCharacterSet characterSetWithCharactersInString:CHANNEL_ACCEPTABLE_CHARACTERS] invertedSet];
        
        NSString *filtered = [[string componentsSeparatedByCharactersInSet:cs] componentsJoinedByString:@""];
        
        return [string isEqualToString:filtered];
    } else if (textField == self.uidTF) {
        NSCharacterSet *cs = [[NSCharacterSet characterSetWithCharactersInString:UID_ACCEPTABLE_CHARACTERS] invertedSet];
        
        NSString *filtered = [[string componentsSeparatedByCharactersInSet:cs] componentsJoinedByString:@""];
        
        return [string isEqualToString:filtered];
    } else {
        return true;
    }
}
- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self resignTextFileds];
}

#pragma mark - helper

- (BOOL)p_isValidSDK {
    if ([EVSDKManager isSDKInitedSuccess] == NO) {
        [[CCAlertManager shareInstance] performComfirmTitle:@"提示" message:@"SDK 尚未初始化" comfirmTitle:@"ok" WithComfirm:nil];
        return NO;
    }
    
    return YES;
}

- (BOOL)p_isValidChannel {
    if (self.channelTF.text.length == 0 || self.channelTF.text == nil) {
        [[CCAlertManager shareInstance] performComfirmTitle:@"提示" message:@"请输入频道名" comfirmTitle:@"好的" WithComfirm:nil];
        return NO;
    }
    
    return YES;
}

- (void)resignTextFileds {
    [self.channelTF resignFirstResponder];
    [self.uidTF resignFirstResponder];
}

@end

File: EasyILive_Swift/EVRTCRootViewController.swift
//
//  EVRTCRootViewController.swift
//  EasyILive_iOS
//
//  Created by Lcrnice on 2017/8/7.
//  Copyright © 2017年 Easyvaas. All rights reserved.
//

import UIKit
import EVSDKBaseFramework
import EVRTCFramework

class EVRTCRootViewController: UIViewController {
  
  @IBOutlet weak var channelTF: UITextField!
  @IBOutlet weak var uidTF: UITextField!
  @IBOutlet weak var versionLabel: UILabel!
  
  fileprivate var currentProfile: EVRtcVideoProfile = ._640x360
  static let kShowRTC = "ShowRTC"
  static let kEasyvaas = URL(string: "http://easyvaas.com")
  
  deinit {
    NotificationCenter.default.removeObserver(self)
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    // Do any additional setup after loading the view.
    
    self.addEVObserver()
    EVSDKManager.initSDK(withAppID: "", appKey: "", appSecret: "", userID: "rtcSwiftTester")
    versionLabel.text = String(format: "Version:%@", EVSDKManager.sdkVersion())
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
    // Dispose of any resources that can be recreated.
  }
  
  
  // MARK: - Navigation
  override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
    // Get the new view controller using segue.destinationViewController.
    // Pass the selected object to the new view controller.
    
    resignTextFileds()
    
    if EVRTCRootViewController.kShowRTC == segue.identifier {
      let rtcVC = segue.destination as! EVRTCViewController
      rtcVC.roomName = channelTF.text!
      if uidTF.text?.isEmpty == false {
        rtcVC.uidString = uidTF.text!
      }
      rtcVC.role = sender as! EVRtcClientRole
      rtcVC.profile = currentProfile
    }
    
    if segue.destination is UINavigationController {
      let nav = segue.destination as! UINavigationController
      let settingVC = nav.viewControllers.first as! EVRTCSettingViewController
      settingVC.currentProfile = self.currentProfile.rawValue
    }
  }
  
  
  @IBAction func swiftUnwindSegueToRootVC(segue: UIStoryboardSegue) {
    self.dismiss(animated: true, completion: nil)
  }
  
  @IBAction func swiftUnwindSegueConfigToRootVC(segue: UIStoryboardSegue) {
    if segue.source is EVRTCSettingViewController {
      let sourceVC = segue.source as! EVRTCSettingViewController
      currentProfile = EVRtcVideoProfile(rawValue: sourceVC.currentProfile)!
    }
    self.dismiss(animated: true, completion: nil)
  }
  
  // MARK: - Outlet actions
  @IBAction func scanQR(_ sender: Any) {
    let scanVC: EVScanQRViewController = EVScanQRViewController()
    scanVC.getQrCode = { [weak self] qr in
      guard let sSelf = self, let qrString = qr else {
        return
      }
      
      sSelf.channelTF.text = qrString
    }
    self.present(scanVC, animated: true, completion: nil)
  }
  
  @IBAction func joinChannel(_ sender: Any) {
    guard isSDKValid(), isChannelValid() else {
      return
    }
    
    EVMediaAuth.checkAndRequestMicPhoneAndCameraUserAuthed({ [weak self] in
      DispatchQueue.main.async {
        self?.showDetailVC(role: .clientRole_LiveGuest)
      }
    }, userDeny: nil)
  }
  
  @IBAction func createChannel(_ sender: Any) {
    guard isSDKValid() else {
      return
    }
    
    EVMediaAuth.checkAndRequestMicPhoneAndCameraUserAuthed({ [weak self] in
      DispatchQueue.main.async {
        self?.showDetailVC(role: .clientRole_Master)
      }
      }, userDeny: nil)
  }
  
  @IBAction func watchLivingChannel(_ sender: Any) {
    guard isSDKValid(), isChannelValid() else {
      return
    }
    
    self.showDetailVC(role: .clientRole_Guest)
  }
  
  @IBAction func easyvaas(_ sender: Any) {
    if #available(iOS 10.0, *) {
      UIApplication.shared.open(EVRTCRootViewController.kEasyvaas!, options: [:], completionHandler: { (success) in
      })
    } else {
      UIApplication.shared.openURL(EVRTCRootViewController.kEasyvaas!)
    }
  }
  
}


// MARK: - 注册SDK
extension EVRTCRootViewController {
  fileprivate func addEVObserver() {
    NotificationCenter.default.addObserver(forName: .EVSDKInitSuccess, object: nil, queue: OperationQueue.main, using: { _ in
      print("SDK 初始化成功")
    })
    
    NotificationCenter.default.addObserver(forName: .EVSDKInitError, object: nil, queue: OperationQueue.main, using: { notify in
      print("SDK 初始化失败")
      CCAlertManager.shareInstance().performComfirmTitle(notify.name.rawValue, message: String(format: "%@", notify.object as! CVarArg), comfirmTitle: "ok", withComfirm: nil)
    })
  }
}

// MARK: - Helpers
extension EVRTCRootViewController {
  fileprivate func isSDKValid() -> Bool {
    if EVSDKManager.isSDKInitedSuccess() {
      return true
    }
    
    CCAlertManager.shareInstance().performComfirmTitle("提示", message: "SDK 尚未初始化", comfirmTitle: "ok", withComfirm: nil)
    return false
  }
  
  fileprivate func isChannelValid() -> Bool {
    if (channelTF.text?.isEmpty)! || channelTF.text == nil {
      CCAlertManager.shareInstance().performComfirmTitle("提示", message: "请输入频道名", comfirmTitle: "ok", withComfirm: nil)
      return false
    }
    
    return true
  }
  
  fileprivate func showDetailVC(role currentRole:EVRtcClientRole) {
    self.performSegue(withIdentifier: EVRTCRootViewController.kShowRTC, sender: currentRole)
  }
  
  fileprivate func resignTextFileds() {
    channelTF.resignFirstResponder()
    uidTF.resignFirstResponder()
  }
}

